### ISSUE TITLE
Need Help

### ISSUE THREAD

**Issue by fire-keeper** [lin-shuyu/VAE-LSTM-for-anomaly-detection#1]
Hello , I have read your paper --- Anomaly Detection for Time Series Using VAE-LSTM Hybrid Model . I am curious about the exact structure of your model and I am doing some research on Anomaly Detection for Time Series , so I click the link you leave on your paper . However , I find nothing but a readme in the repository .
If you have no spare time to post your code, could you please tell me the detail of  your model's structure.
Best wishes to you.

**Comment by lin-shuyu** [comment]
Hi Fire-keeper,

Many thanks for getting in touch and going through our paper! Apologies on not seeing your question earlier - I'm still new to releasing code on github. 

Our VAE model uses simple CNN operations. We treat the time series as a 2D signal, where the feature channel is the 2nd spatial dimension. This would allow the model to learn convolutional filters that capture correlations across different channel dimensions, if your time series contains multiple channel. 

We just released our code today! So detailed architecture of our model can be found in models.py, which is positioned in this repository under code/ subfolder. 

Hope you find our paper and codes useful. Looking forward to seeing your anomaly detection algorithm soon! 

Best wishes,
Lin

**Comment by haopo2005** [comment]
Hi,
if there are multiple channel for my time series, then there is no need to expand the feature?
input_tensor = tf.expand_dims(self.original_signal, -1)

Will it be different result between multiple conv over each signal and single conv over multiple signal ?

**Comment by lin-shuyu** [comment]
Hi Haopo, 

Apologies on the delay in getting back! 

I have tried this method with multi-channel time series. Yes, in this case, you don't need to expand_dims of the original_signal. 

We use conv2d because we want to take care of the cross-channel dependency / correlations in the conv kernels. I think the only thing you need to change is to remove expand_dims, as you spotted above. Also, you will need to change the "n_channel" in the config.jason file to the number of channels in your dataset. 

You probably need to process your dataset in the format we did with NAB dataset. Sorry the dataloader.py probably was not the most well documented code.

Hope this helps! 

Best wishes,
Lin

**Comment by miemie030205** [comment]
Hello? I'm very sorry to disturb you. I'm an undergraduate student and I'm reproducing your code -VAE-LSTM-for-anomaly-detection. For writing my thesis, I have several questions to ask you.
1) The data of the VAE model in the comparative experiments you mentioned in your paper, was it someone else's code that you reproduced or was it written by yourself? In the detection code of your code - nab-anomalies detection.ipynb, I saw a piece of code that only used the VAE model for comparison. I encapsulated it into a function, corresponding to the output of values such as F1. May I ask the value of the VAE model in your paper? Was it obtained through the code in your code that only uses the VAE model?
2) May I ask how you choose the window? I have found that by randomly expanding the window, the value of F1 can become very large. May I ask what the logic is for you to choose the window size? How did you choose the Windows for these five datasets? What size Windows did you respectively choose for these five datasets?
3) I saw two methods for calculating the value of F1 in your code, one before enhancement and the other after enhancement. May I ask if the data presented in your paper is the value before enhancement or the value after enhancement?
4) May I ask if all the five datasets you selected are point exceptions, without context exceptions or collective exceptions? Is there a distinction between short-term exceptions and long-term exceptions in your dataset?
5) May I ask how you obtained the code for the models other than the one you wrote for the comparative experiments in your paper? I checked your reference, but I didn't see where the code was.
I'm very sorry to disturb you. Looking forward to your reply.
